### ISSUE TITLE
Provide a server side transaction hook

### ISSUE THREAD

**Issue by phreed** [webgme/webgme#1627]
Enhancement.
Certain transactions update the data store and are shared with other user instances.
As these transactions arrive at the server provide a hook whereby they may be intercepted and shared with another principal.
These other principals could include: alternate databases with their own schema; durable message queues like Kafka; scalability protocols like 0mq; IDE like eclipse or ideas; ...
In addition to writing transaction messages the hook should also allow for reading transactions as if originating from the webGME UI.

An added benefit would be in the testing space as it would allow for the recording a playback of transactions.

An important consideration is in the specification of transaction messages as there schema would need to be part of the webGME API.

**Comment by phreed** [comment]
Yes, webhooks could be used to capture the browser generated commits.
What would the mechanism be for injecting commits, e.g. replay recorded commits?

**Comment by pmeijer** [comment]
That would be the project API, accessible either through the [REST API (does not support insertion of objects though)](https://editor.webgme.org/developer/api#projects) or websockets via a [client in javascript*](https://editor.webgme.org/docs/source/Project.html).
However in order to commit any changes that are related to the model-tree - you need to core API to generate the data-blobs inserted with the project API.. The core and project API is practically what the plugin API is built up of. And these can be invoked from command line using [run_plugin.js](https://github.com/webgme/webgme-engine/blob/master/src/bin/run_plugin.js#L61).

* Both the core and project api have bindings in [python](https://github.com/webgme/bindings)

**Comment by phreed** [comment]
What would be an example of an object that could not be inserted via REST API?
These objects could be inserted using a websocket client?

**Comment by pmeijer** [comment]
That would be [commit-objects](https://editor.webgme.org/api/projects/demo/FiniteStateMachine/commits?n=1) and [core data-blobs](https://editor.webgme.org/api/projects/demo/FiniteStateMachine/branches/master/tree/)

There's nothing inherent that restricts the REST API from exposing such end-points though.

**Comment by phreed** [comment]
I would rather use a websocket. 
Can the objects be inserted using a websocket client?

**Comment by pmeijer** [comment]
[Project.makeCommit](https://editor.webgme.org/docs/source/Project.html#makeCommit__anchor).
The commit-object is created (and later inserted) as part of invoking makeCommit. The core-data-blobs are generated at [Core.persist](https://editor.webgme.org/docs/source/Core.html#persist__anchor). In later versions these objects are actually diffs generated based on the changes in the model-tree - however the project API also accepts the complete objects. 

The [save-method in the plugin-base](https://github.com/webgme/webgme-engine/blob/master/src/plugin/PluginBase.js#L461-L471) is a good example where this takes place.
